src/Discord.Net.SlashCommands/SlashCommandService.cs

```diff
@@ -0,0 +1,675 @@
+                }
+                return moduleDefs.Values;
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+        /// <summary>
+        /// Add a command module manually to the command service
+        /// </summary>
+        /// <typeparam name="T">Type of the module</typeparam>
+        /// <param name="services">Service provider that will be used to build this module</param>
+        /// <returns>A task representing the module loading process</returns>
+        /// <exception cref="ArgumentException">Thrown when a module that is already present in the command service is trying to be added</exception>
+        /// <exception cref="InvalidOperationException">Thrown when the <typeparamref name="T"/> is not a valid module definition</exception>
+        public async Task AddModuleAsync<T> (IServiceProvider services)
+        {
+            if (!typeof(ISlashModuleBase).IsAssignableFrom(typeof(T)))
+                throw new ArgumentException("Type parameter must be a type of Slash Module", "T");
+
+            services = services ?? EmptyServiceProvider.Instance;
+
+            await _lock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                var typeInfo = typeof(T).GetTypeInfo();
+
+                if (_typedModuleDefs.ContainsKey(typeInfo))
+                    throw new ArgumentException("Module definition for this type already exists.");
+
+                var moduleDef = await ModuleClassBuilder.BuildAsync(new List<TypeInfo> { typeof(T).GetTypeInfo() }, this, services).ConfigureAwait(false);
+
+                if (moduleDef[typeof(T)] == default(ModuleInfo))
+                    throw new InvalidOperationException($"Could not build the module {typeInfo.FullName}, did you pass an invalid type?");
+
+                if (!_typedModuleDefs.TryAdd(typeof(T), moduleDef[typeof(T)]))
+                    throw new ArgumentException("Module definition for this type already exists.");
+
+                LoadModuleInternal(moduleDef[typeof(T)]);
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+        /// <summary>
+        /// Register/update the Application Commands to Discord
+        /// </summary>
+        /// <param name="guildId">The Id that belongs to the guild, the commands will be registered to</param>
+        /// <param name="deleteMissing">If true, delete all of the commands that are not registered in the <see cref="SlashCommandService"/></param>
+        /// <returns>A task representing the command registration process, with a collection of <see cref="RestGuildCommand"/> containing the
+        /// commands that are currently registered to the provided guild as its result.
+        /// </returns>
+        public async Task<IReadOnlyCollection<RestGuildCommand>> RegisterCommandsToGuildAsync (ulong guildId, bool deleteMissing = true)
```

Comment:
I find this redundant, Commands should be registered in factory fashion. When the service is created all command defs should be built with their corresponding properties then should be created automatically instead of registering them with a function call.

Somthing to add to this; you should be able to add/remove certain command to and from guilds at runtime.